### PR TITLE
Telemetry: Metric collection from all resources

### PR DESCRIPTION
**Pull request by NajmudheenCT**
**What this PR does / why we need it**:
Earlier metric driver was designed to collect for one instance at a time. With these approach we can collect only instances available in etcd (hotpot DB) . Driver interface change is required to collect for all resources. This change set contains
1. Driver interface change to collect All resource metrics supported.
2. Lvm driver change to discover LVM resources and collect metrics for all.
3. LVM metrics  mock test changes
4.POST API changes to incorporate driver interface change. Now only Driver type is passed as parameter in POST request body.
4. Controller logic change for picking the dock type.Now the controller list all available docks and select the one which matches driver type in POST request body.
5. lvm exporter change according to the new driver interface.

**Which issue this PR fixes** 
https://github.com/opensds/opensds/issues/748
https://github.com/opensds/opensds/issues/747

### CODE REVIEW

**Comment by NajmudheenCT** [comment]
@skdwriting @leonwanghui @xing-yang @rhsakarpos

**Comment by NajmudheenCT** [comment]
> Please use `make goimports` to format the code.

Thanks @leonwanghui , updated , please check

**Comment by leonwanghui** [comment]
It seems like you close this PR?

**Comment by NajmudheenCT** [comment]
> It seems like you close this PR?

Sorry, that was by mistake. I reopened it

File: contrib/drivers/lvm/lvm_metrics_test.go [hunk 1]
```diff
@@ -135,9 +178,9 @@ func TestCollectMetrics(t *testing.T) {
 		}
 	}
 	if !b {
-		t.Errorf("expected metric spec")
+		fmt.Printf("expected metric spec")
```
Comment (leonwanghui) [inline contrib/drivers/lvm/lvm_metrics_test.go]: Suggest `t.Log()`.

File: contrib/drivers/lvm/lvm_metrics_test.go [hunk 2]
```diff
@@ -135,9 +178,9 @@ func TestCollectMetrics(t *testing.T) {
...
 		printMetricSpec(expectedMetrics)
-		t.Errorf("returned metric spec")
+		fmt.Printf("returned metric spec")
```
Comment (leonwanghui) [inline contrib/drivers/lvm/lvm_metrics_test.go]: Same here.

**Comment by leonwanghui** [comment]
@xing-yang @wisererik PTAL, thanks!

File: contrib/drivers/ceph/ceph_metrics.go [hunk 3]
```diff
@@ -101,9 +101,12 @@ func (d *MetricDriver) ValidateMetricsSupportList(metricList []string, resourceT
 //	metricsList-> posted metric list
 //	instanceID -> posted instanceID
 //	metricArray	-> the array of metrics to be returned
-func (d *MetricDriver) CollectMetrics(metricsList []string, instanceID string) ([]*model.MetricSpec, error) {
+func (d *MetricDriver) CollectMetrics() ([]*model.MetricSpec, error) {
 
 	//validate metric support list
+	// Todo: Remove this two lines when Ceph driver implements modified driver interface
+	metricsList := []string{"pool_used_bytes", "pool_raw_used_bytes", "pool_available_bytes", "pool_objects_total", "pool_dirty_objects_total", "pool_read_total", "pool_read_bytes_total", "pool_write_total", "pool_write_bytes_total"}
+	instanceID := "pool1"
```
Comment (xing-yang) [inline contrib/drivers/ceph/ceph_metrics.go]: Why is this hard-coded to "pool1"?

File: pkg/controller/controller.go [hunk 4]
```diff
@@ -1108,31 +1108,32 @@ func (c *Controller) CollectMetrics(context context.Context, opt *pb.CollectMetr
...
-		log.Errorf("get volume by id %s failed in CollectMetrics method: %s", opt.InstanceId, err.Error())
+		log.Errorf("list dock  failed in CollectMetrics method: %s", err.Error())
 		return pb.GenericResponseError(err), err
 	}
+	for i, d := range dockSpec {
+		if d.DriverName == opt.DriverName {
+			log.Infof("driver found driver: %s", d.DriverName)
+			dockInfo, err := db.C.GetDock(ctx, dockSpec[i].BaseModel.Id)
+			if err != nil {
+				//log.Errorf("error %s when search dock in db by pool id: %s", err.Error(), vol.PoolId)
```
Comment (xing-yang) [inline pkg/controller/controller.go]: Remove commented out code.

File: contrib/drivers/lvm/lvm_metrics.go [hunk 5]
```diff
@@ -127,67 +117,109 @@ func (d *MetricDriver) ValidateMetricsSupportList(metricList []string, resourceT
...
-		tempMetricArray = append(tempMetricArray, metric)
 	}
 	metricArray := tempMetricArray
 	return metricArray, err
 }
+func convert(instanceID string) string {
+	// systat utilities (sar/iostat) returnes  volume with -- instead of -, so we need to modify volume name to map lvs output
+	instanceID = strings.Replace(instanceID, "-", "--", -1)
+	//add opensds--volumes--default-- to the start of volume
+	instanceID = "opensds--volumes--default-" + instanceID
```
Comment (rhsakarpos) [inline contrib/drivers/lvm/lvm_metrics.go]: please check, default-- or default-

File: contrib/drivers/lvm/lvm_metrics.go [hunk 6]
```diff
@@ -127,67 +117,109 @@ func (d *MetricDriver) ValidateMetricsSupportList(metricList []string, resourceT
...
+	return instanceID
+}
+func formatDiskName(instanceID string) string {
+	// systat(sar/iostat) returns only disk name. We need to add /dev/ to match with pvs output
+	instanceID = strings.Replace(instanceID, "/dev/", "", -1)
+	//add opensds--volumes--default-- to the start of volume
```
Comment (rhsakarpos) [inline contrib/drivers/lvm/lvm_metrics.go]: not needed here, please remove

File: contrib/drivers/ceph/ceph_metrics.go (hunk 3 above)
Comment (NajmudheenCT) [reply contrib/drivers/ceph/ceph_metrics.go]: This is a substitute to make the build  successful after driver interface change. C2C team will update this code in comply with new driver interface .

File: contrib/drivers/lvm/lvm_metrics.go (hunk 5 above)
Comment (NajmudheenCT) [reply contrib/drivers/lvm/lvm_metrics.go]: Checked , Its single '-' only fin that place

File: contrib/drivers/lvm/lvm_metrics.go (hunk 6 above)
Comment (NajmudheenCT) [reply contrib/drivers/lvm/lvm_metrics.go]: removed, Thanks

File: pkg/controller/controller.go (hunk 4 above)
Comment (NajmudheenCT) [reply pkg/controller/controller.go]: Modified , Thanks

File: contrib/drivers/lvm/lvm_metrics.go [hunk 7]
```diff
@@ -127,67 +117,107 @@ func (d *MetricDriver) ValidateMetricsSupportList(metricList []string, resourceT
...
 	if supportedMetrics == nil {
 		log.Infof("no metrics found in the  supported metric list")
 	}
-	metricMap, labelMap, err := d.cli.CollectMetrics(supportedMetrics, instanceID)
-
+	// discover lvm volumes
+	volumeList, err := d.cli.DiscoverVolumes()
+	// discover lvm physical volumes
+	DiskList, err := d.cli.DiscoverDisks()
+	metricMap, labelMap, err := d.cli.CollectMetrics(supportedMetrics)
```
Comment (leonwanghui) [inline contrib/drivers/lvm/lvm_metrics.go]: This err will override the last two err variables before, if you just want to validate this error, please use `_` instead for last two.

Comment (NajmudheenCT) [reply contrib/drivers/lvm/lvm_metrics.go]: > This err will override the last two err variables before, if you just want to validate this error, please use `_` instead for last two.

Thanks for pinpoint, I have handled this

**Comment by leonwanghui** [comment]
@wisererik PTAL

**Pull request merged**